README.md
# MRU-Tag
  *Spend less time navigating and more time coding!*

## Introduction
MRU-Tag allows you to quickly jump to your most recently used files and tags(functions/methods/etc).  

![Sample Output](https://github.com/sequenceGeek/MRU-Function/raw/master/sample.gif)

## Install
(Note: MRU-Tag currently requires Python to be installed and has not been tested on Windows machines)

1. Install Exuberant Ctags.

        sudo apt-get install exuberant-ctags  
    
2. Install [Pathogen](https://github.com/tpope/vim-pathogen) (Vim Plugin Manager).  
3. Clone MRU-Function repo into your vim bundle directory.  

        cd ~/.vim/bundle  
        git clone git@github.com:sequenceGeek/MRU-Function.git

## Usage
(Note: start editing code inside functions/methods/etc to populate browser)
- Pressing `<F3>` opens the MRU Tag Browser.
- Select a file and press `<Enter>` to expand the file's MRU tags.
- Select a tag and press `<Enter>` to jump to the tag.


CHANGELOG.md

## Todo/Issues
- [ ] Include (in the browser) opened files even when no code editing occurs.
- [ ] Python 3 compatible
- [ ] Add "pick function by number"
- [ ] Windows compatible

### 2013-10-27
- [x] Add gif to show usage.
- [x] Updated browser formatting and syntax highlighting

### 2013-10-26
- [x] Syntax/Colors for Browser 
- [x] Custom "nearest tag function" for python to handle nested functions/methods.

### 2013-10-20
- [x] Expand Menu Display (by file, and function)
- [x] Refactored

### 2013-10-17
- [x] Currently modified buffers show/don't show tags incorrectly
- [x] Changed Folder Structure

### 2013-10-16
- [x] Fix "wrongly logged tag due to file being edited" bug.
- [x] Sorting is now done correctly
- [x] Won't re-open/edit same file if you go to function in same file


plugin/tests/mockFile1.py
import os

def func_outside():    
    ''' blah blah doc blah blah'''

    if "hello" == "world":
        return None

    def func_inside():
        
        while True:
            print "hello"

    #test outside
    if "hello":
        print "nothing"



plugin/tags.py
import subprocess
import json
import os
import re

from mruutility import first_non_whitespace_index

PLUGIN_HOME = os.path.dirname(os.path.abspath(__file__))

class TagFile:

    def __init__(self, srcFileName):
        '''create tags for specific file'''
        
        self.srcFileName = srcFileName
        self.tags = []
        self.get_tags_in_file()

    def get_tags_in_file(self):

        [self.tags.append(Tag(ctagsLine = line)) for line in tagify_file(self.srcFileName)]

def tagify_file(fileName):
    '''execute ctags on a specific file (with linenumber option and get the output'''

    # have to add the vimL mapping to prevent auto-sourcing from Pathogen
    ctagProcess = subprocess.Popen(["ctags-exuberant", "--fields=+n", "-f", "-", fileName], stdout=subprocess.PIPE) 
    fOutput = ctagProcess.stdout.readlines()

    return fOutput

class Tag:

    def __init__(self, ctagsLine = None, jsonLine = None):

        self.name = None
        self.srcName = None
        self.regex = None
        self.type = None
        self.lineNumber = None #1-based
        self.prototype = None
        self.ctagsLine = ctagsLine
        self.jsonLine = jsonLine
        self.fnwi = None

        #cursor location information (should re-factor)
        self.cursorOffsetLine = None
        self.cursorColumnNumber = None

        # optional initialization
        if ctagsLine:
            self.parse_ctags_output_line()
        elif jsonLine:
            self.from_json()
       
        # first character is always "^"
        self.fnwi = first_non_whitespace_index(self.regex[2:])

    def __repr__(self):
        return "\t".join([str(x) for x in (self.name, self.srcName, self.type)])

    def function_choice_output(self):
        
        if self.prototype:
            tName = self.name + "(" + self.prototype.replace("class:", "") + ")"
        else:
            tName = self.name

        whitespace = ' ' * 1000
        outText = tName + whitespace + '\t' + os.path.basename(self.srcName) + "\t" + str(self.cursorOffsetLine) + "\t" + str(self.cursorColumnNumber) + "\t" + self.regex.replace(r'/;"', "") + "\t" + self.srcName
        return outText

    def check_existence(self, cBufferFN):
        '''check if tag exists in file'''
      
        checkFN = self.srcName
        if cBufferFN == self.srcName:
            # search through the modified buffer instead
            checkFN = PLUGIN_HOME + "/../tmp/cBuffers/cBuffer." + self.srcName.split(".")[-1]

        with open(checkFN, 'r') as f: 
            reg = self.truncate_regex()

            # why do I not have to escape these in vim? @wtodo @frequent-function @python 
            reg = reg.replace("(", "\(")
            reg = reg.replace(")", "\)")
            pat = re.compile(reg, re.MULTILINE)
            # existence of a regex in string @example @regex @python 
            if pat.search(f.read()): 
                return True
            else:
                return False

    def truncate_regex(self):

        tRegex = self.regex
        if self.regex[0] == "/":
            tRegex = self.regex[1:] 

        tRegex = tRegex.replace(r'/;"', "")
        return tRegex

    def parse_ctags_output_line(self):

        ls = self.ctagsLine.strip().split("\t")
        self.name = ls[0]
        self.srcName = ls[1]
        self.regex = ls[2]
        self.type = ls[3]
        self.lineNumber = int(ls[4].replace("line:", ""))
        try:
            self.prototype = ls[5]
        except IndexError:
            self.prototype = None 

    def to_json(self):
        # make a json dict maker @wtodo @python 
        jDict = {"name": self.name,
                 "srcName": self.srcName,
                 "regex": self.regex,
                 "type": self.type,
                 "lineNumber": self.lineNumber,
                 "prototype": self.prototype,
                 "cursorOffsetLine": self.cursorOffsetLine,
                 "cursorColumnNumber": self.cursorColumnNumber
                 }

        return json.dumps(jDict)

    def from_json(self):
        objProperties = json.loads(self.jsonLine.strip()) 
        for attName, attProp in objProperties.items():
            setattr(self, attName, attProp)

def test_tagify(srcFileName):
    '''just testing if ctagging works'''
    tFile = TagFile(srcFileName)
    print tFile
    for tag in tFile.tags:
        print tag.name
        print tag.lineNumber
        print tag.regex[2:]
        print tag.fnwi

if __name__ == "__main__":
    import sys
    assert sys.argv[1] in globals(), "Need name of fxn to run from command line!"
    fxnToRun = globals()[sys.argv[1]] 
    fxnToRun(*sys.argv[2:])



plugin/cgtools.py


def flatten(l, ltypes=(list, tuple)):
    '''Flatten multi-level lists of lists
    Credit: http://rightfootin.blogspot.com/2006/09/more-on-python-flatten.html
   
    @flatten @python @list @example
    USAGE
    from cgtools import flatten

    a = [[1,2,3], [4,5,6]]
    allAs = flatten(a)
    '''

    ltype = type(l)
    l = list(l)
    i = 0
    while i < len(l):
        while isinstance(l[i], ltypes):
            if not l[i]:
                l.pop(i)
                i -= 1
                break
            else:
                l[i:i + 1] = l[i]
        i += 1
    return ltype(l)

def groupby(listOfObjects, attributeName):
    '''
    group list of object by a property of the object 
    properties of each object must be of the same type 
    lists must not include complex types (no list of lists)
    '''

   
    allowableTypes = [type(x) for x in [1,'a', True, None, tuple(), list()]]

    attribute_objects = {} 
    for obj in listOfObjects:
        attributeValue = getattr(obj, attributeName) # @python @getattr @attribute @example   
        if type(attributeValue) not in allowableTypes:
            raise NameError("Attribute must be simple type or tuple") # custom error @wtodo @programming

        if type(attributeValue) == type(list()):
            attributeValue = tuple(sorted(attributeValue))

        attribute_objects.setdefault(attributeValue, []).append(obj)

    return attribute_objects

def groupbytraverse(listOfObjects, attributeName):
    '''
    group list of objects by the list elements of specified attributeName
    lists must be 1 dimensional
    
    Make an example @wtodo @cgtools @programming
    '''

    allowableTypesTraverse = [type(x) for x in [list(), tuple()]]
    allowableTypesSimple = [type(x) for x in [1,'a', True, None]]

    attribute_objects = {} 
    for obj in listOfObjects:
        attributeValue = getattr(obj, attributeName) 
        if type(attributeValue) not in allowableTypesTraverse:
            raise NameError("Attribute value type must be list or tuple")

        for element in attributeValue:
            if type(element) not in allowableTypesSimple:
                raise NameError("List element type must be simple")

            attribute_objects.setdefault(element, []).append(obj)

    return attribute_objects


plugin/recentfxn.py
#!/usr/bin/env python

from __future__ import print_function
import os
import collections
import logging

from mruutility import groupby
from expandmenu import ExpandMenu
from tags import TagFile, Tag
from pythontag import py_get_nearest_tag

# get directory of this file @i @python
PLUGIN_HOME = os.path.dirname(os.path.abspath(__file__))
LOG_FILE = PLUGIN_HOME + "/../data/locations.log"
WINDOW_TEXT = PLUGIN_HOME + "/../tmp/windowtext.txt"
MENU_STATE = PLUGIN_HOME + "/../tmp/menu_state.txt"

def log_tag_info(srcFile, lineNum, colNum, fileExtension, logFile):
    ''' Log nearest tag when vim sends log signal '''
    #get the nearest tag, will return None if no tag match
    cBufferFN = PLUGIN_HOME + "/../tmp/cBuffers/cBuffer." + fileExtension
    logging.info("cbuffer file: " + cBufferFN)
    
    logging.info("getting nearest tag... " + cBufferFN)
    nTag = get_nearby_tag(cBufferFN, lineNum, colNum, fileExtension)
 
    if nTag:
        logging.info("FOUND TAG: %s", nTag.name)

        #change the file name to not point to the temp file
        nTag.srcName = srcFile
        with open(logFile, 'a') as f:
            f.write(nTag.to_json() + "\n")
    else:
        logging.warning("NO TAGS")

def get_nearby_tag(srcFile, lineNum, colNum, fileExtension, tagTypes = "fm"):
    ''' f = function, m = method '''
    lineNum, colNum = int(lineNum), int(colNum)
    tFile = TagFile(srcFile)
    lineNum_tag = groupby(tFile.tags, "lineNumber")

    # custom nearest tag functions
    logging.info("src, fe, line: %s, %s, %d", srcFile, fileExtension, lineNum)
    if fileExtension == "py":
        logging.info("PYTHON tag... ")
        (nTag, tagLineNum) = py_nearest_tag(srcFile, lineNum, colNum,  lineNum_tag, tagTypes)
    else:
        logging.info("GENERAL tag... ")
        (nTag, tagLineNum) = general_nearest_tag(lineNum, lineNum_tag, tagTypes)

    if nTag:
        nTag.cursorOffsetLine = lineNum - tagLineNum
        nTag.cursorColumnNumber = colNum
        return nTag
    else:
        return None

def py_nearest_tag(srcFile, lineNum, colNum, lineNum_tag, tagTypes):

    '''check 1-based line'''
    with open(srcFile, 'r') as f:
        fileText = f.read()

    pyNearest = py_get_nearest_tag(fileText, lineNum, colNum) 
    if pyNearest:
        logging.info("pyNearest: %d %s", pyNearest[0], pyNearest[1])
        tLineNum, tName = pyNearest
        if tLineNum in lineNum_tag:
            nTag = lineNum_tag[tLineNum][0]
            if nTag.type in tagTypes:
                logging.info("FOUND TAG [%s %d]", nTag.name, tLineNum)
                return (nTag, tLineNum)

    # no tags found
    logging.info("NO TAG FOUND")
    return (None, None)

def general_nearest_tag(lineNum, lineNum_tag, tagTypes):
    
    for i in xrange(lineNum, 0, -1):
        if i in lineNum_tag:
            nearTag = lineNum_tag[i][0]
            if nearTag.type in tagTypes:
                return (nearTag, i)

    # no tags
    return (None, None)

def update_log(logFN, cBufferFN):
    '''remove non-existing tags (file changed/deleted or refactored)
    and remove duplicate functions'''
    
    # re-factor log to class @wtodo @frequent-function
    tags = get_unique_tags(load_tags_from_log(logFN))

    # does the file even exist anymore?
    tags = [tag for tag in tags if os.path.exists(tag.srcName)]

    # debug level @programming @wtodo

    # does the function exist anymore, has it been refactored?
    tags = [tag for tag in tags if tag.check_existence(cBufferFN)] 

    with open(logFN, 'w') as f:
        [f.write(tag.to_json() + "\n") for tag in tags]

def get_unique_tags(tags):
    '''tags are returned in order given
    [a,b,c] -> [a,b] (if c was not unique)'''

    uniqueProps = set()
    uniqueTags = []
    while True:
        if not tags:
            break

        currTag = tags.pop()
        tagUniqueProps = (currTag.name, currTag.srcName, currTag.prototype)
        if tagUniqueProps not in uniqueProps:
            uniqueTags.append(currTag)
            uniqueProps.add(tagUniqueProps)

    # popping and appending reversed the list order
    uniqueTags.reverse()

    return uniqueTags

def load_tags_from_log(logFN):

    tags = []
    with open(logFN, 'r') as f:
        for jsonline in f:
            tags.append(Tag(jsonLine = jsonline))

    return tags

def generate_mru_browser_text(cBufferName, browserMode):

    # remove dups and non-existing first
    update_log(LOG_FILE, cBufferName)

    if browserMode == "fxn":
        menu_functions_only()
    elif browserMode == "byfile":
        menu_by_file()
    else:
        raise ValueError("Need Menu Mode")

def menu_by_file():
    '''create an expandable menu
    save text in windowtext file
    save state'''

    menuLines = create_by_file_expand_lines()
    eMenu = ExpandMenu(menuLines = menuLines)
    eMenu.output_menu(WINDOW_TEXT)
    eMenu.save_to_file(MENU_STATE)

def menu_functions_only():
    
    # get resulting tags and display in reverse order (log has newest at end)
    logTags = load_tags_from_log(LOG_FILE)
    logTags.reverse()
    
    with open(WINDOW_TEXT, 'w') as f:
        for tag in logTags:
            f.write(tag.function_choice_output() + "\n")

def create_by_file_expand_lines():
    '''should give this an ordered dict and have it make it'''
    
    # get resulting tags and display in reverse order (log has newest at end)
    logTags = load_tags_from_log(LOG_FILE)
    logTags.reverse()

    # group by the file names, order by mru file
    srcName_tags = collections.OrderedDict()
    for tag in logTags:
        srcName_tags.setdefault(tag.srcName, []).append(tag)

    return srcName_tags

def handle_expand_choice(choice):

    eMenu = ExpandMenu()
    eMenu.load_from_file(MENU_STATE)
    eMenu.handle_expansion_change(choice)
    eMenu.output_menu(WINDOW_TEXT)
    eMenu.save_to_file(MENU_STATE)

if __name__ == "__main__":
    import sys

    # Set up logging
    logging.basicConfig(filename=PLUGIN_HOME + '/../debug.txt', level=logging.INFO)

    # Handle events
    event = sys.argv[1]
    logging.info("***     MAIN     ***")
    logging.info("***     %s     ***", event)
    logging.info("args: %s", sys.argv[1:])
    if event == "log":
        log_tag_info(sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], LOG_FILE)
    elif event == "menu":
        cBuffer, mode = sys.argv[2:]
        generate_mru_browser_text(cBuffer, mode)
    elif event == "expand":
        expand_choice = sys.argv[2]
        handle_expand_choice(expand_choice)
    else:
        raise KeyError("Must Specify Event")     


plugin/expandmenu.py
import os
import json

class ExpandMenu(object):

    def __init__(self, menuLines = None):
        '''menuLines is an OrderedDict'''
        self.menuLines = menuLines
        self.expanded = []
        self.menuText = ""

        if self.menuLines:
            self.create_menu_text(self.menuLines)

    def create_menu_text(self, menuLines):
        '''1D Menu Only
        menuLines is an ORDERED DICT'''

        prefixSep = "  "
        self.menuText = " MRU-Tag (Files sorted by MRU.  Select file to see MRU tags)\n"
        whitespace = " " * 1000
        for i, fName in enumerate(menuLines):
            baseFileName = os.path.basename(fName)
            self.menuText += "+[" + str(i + 1) + "] " + baseFileName + "\n"
            self.menuText +=  prefixSep + "...go to " + baseFileName + whitespace + "\t" + fName + "\n"
            for i, tag in enumerate(menuLines[fName]):
                self.menuText += prefixSep + tag.function_choice_output() + "\n"

    def menu_lines(self):
        return self.menuText.split("\n")

    def output_menu(self, fName):

        tPrint = False
        eCounter = 0
        with open(fName, 'w') as f:
            for i, mLine in enumerate(self.menu_lines()):
                if is_expandable(mLine):
                    eCounter += 1
                    if eCounter in self.expanded:
                        f.write(mLine.replace("+[", "-[") + "\n")
                        tPrint = True
                    else:
                        f.write(mLine + "\n")
                        tPrint = False
                else:
                    # Title menu
                    if i == 0:
                        f.write(mLine + "\n")
                    # Everything else
                    elif tPrint: 
                        f.write(mLine + "\n")

    def handle_expansion_change(self, choice):

        # choice will be given as X of line in [X]
        choice = int(choice)
        if choice in self.expanded:
            self.expanded.remove(choice)
        else:
            self.expanded.append(choice)


    def load_from_file(self, fName):

        with open(fName, 'r') as f:
            self.from_json(f.readline())

    def save_to_file(self, fName): 

        with open(fName, 'w') as f:
            f.write(self.to_json())

    def to_json(self):
        # make a json dict maker @wtodo @python 
        jDict = {"menuText": self.menuText,
                 "expanded": self.expanded,
                 }

        return json.dumps(jDict)

    def from_json(self, jsonLine):
        objProperties = json.loads(jsonLine.strip()) 
        for attName, attProp in objProperties.items():
            setattr(self, attName, attProp)
    
def is_expandable(text):
    return text.startswith("+[") or text.startswith("-[")


plugin/mruutility.py
import re

def first_non_whitespace_index(lineText):
    '''find the index of the first non-whitespace character on line'''
    pat = re.compile(r'(\s*)(\S)')
    m = pat.match(lineText)
    if m:
        # start(2) gets start of 2nd group
        return m.start(2)
    else:
        return None

def is_comment_line(lineText):
    pat = re.compile(r'\s*(#|\'\'\'|\"\"\")')
    return bool(pat.match(lineText))

def is_blank_line(lineText):
    pat = re.compile(r'\s*\n')
    m = pat.match(lineText)
    return bool(m)

def flatten(l, ltypes=(list, tuple)):
    '''Flatten multi-level lists of lists
    Credit: http://rightfootin.blogspot.com/2006/09/more-on-python-flatten.html
   
    @flatten @python @list @example
    USAGE
    from cgtools import flatten

    a = [[1,2,3], [4,5,6]]
    allAs = flatten(a)
    '''

    ltype = type(l)
    l = list(l)
    i = 0
    while i < len(l):
        while isinstance(l[i], ltypes):
            if not l[i]:
                l.pop(i)
                i -= 1
                break
            else:
                l[i:i + 1] = l[i]
        i += 1
    return ltype(l)

def groupby(listOfObjects, attributeName):
    '''
    group list of object by a property of the object 
    properties of each object must be of the same type 
    lists must not include complex types (no list of lists)
    '''

   
    allowableTypes = [type(x) for x in [1,'a', True, None, tuple(), list()]]

    attribute_objects = {} 
    for obj in listOfObjects:
        attributeValue = getattr(obj, attributeName) # @python @getattr @attribute @example   
        if type(attributeValue) not in allowableTypes:
            raise NameError("Attribute must be simple type or tuple") # custom error @wtodo @programming

        if type(attributeValue) == type(list()):
            attributeValue = tuple(sorted(attributeValue))

        attribute_objects.setdefault(attributeValue, []).append(obj)

    return attribute_objects

def groupbytraverse(listOfObjects, attributeName):
    '''
    group list of objects by the list elements of specified attributeName
    lists must be 1 dimensional
    
    Make an example @wtodo @cgtools @programming
    '''

    allowableTypesTraverse = [type(x) for x in [list(), tuple()]]
    allowableTypesSimple = [type(x) for x in [1,'a', True, None]]

    attribute_objects = {} 
    for obj in listOfObjects:
        attributeValue = getattr(obj, attributeName) 
        if type(attributeValue) not in allowableTypesTraverse:
            raise NameError("Attribute value type must be list or tuple")

        for element in attributeValue:
            if type(element) not in allowableTypesSimple:
                raise NameError("List element type must be simple")

            attribute_objects.setdefault(element, []).append(obj)

    return attribute_objects


tmp/cBuffers/cBuffer.py
#!/usr/bin/env python

import os
import subprocess
import json
import re

import cgtools

# get directory of this file @i @python
PLUGIN_HOME = os.path.dirname(os.path.abspath(__file__))
LOG_FILE = PLUGIN_HOME + "/../data/locations.log"
WINDOW_TEXT = PLUGIN_HOME + "/../tmp/windowtext.txt"
DLOG_FILE = PLUGIN_HOME + "/../debug.txt"

def dlog(outString):
    with open(DLOG_FILE, 'a') as f:
        f.write(str(outString) + "\n")

class TagFile:

    def __init__(self, srcFileName):
        '''create tags for specific file'''
        
        self.srcFileName = srcFileName
        self.tags = []

        [self.tags.append(Tag(ctagsLine = line)) for line in self.tagify_file(srcFileName)]

    def tagify_file(self, fileName):
        '''execute ctags on a specific file (with linenumber option and get the output'''

        # have to add the vimL mapping to prevent auto-sourcing from Pathogen
        ctagProcess = subprocess.Popen(["ctags", "--fields=+n", "-f", "-", fileName], stdout=subprocess.PIPE) 
        fOutput = ctagProcess.stdout.readlines()

        return fOutput

class Tag:

    def __init__(self, ctagsLine = None, jsonLine = None):

        self.name = None
        self.srcName = None
        self.regex = None
        self.type = None
        self.lineNumber = None
        self.prototype = None
        self.ctagsLine = ctagsLine
        self.jsonLine = jsonLine

        #cursor location information (should re-factor)
        self.cursorOffsetLine = None
        self.cursorColumnNumber = None

        # optional initialization
        if ctagsLine:
            self.parse_ctags_output_line()
        elif jsonLine:
            self.from_json()

    def __repr__(self):
        return "\t".join([str(x) for x in (self.name, self.srcName, self.type)])

    def check_existence(self, cBufferFN):
      
        dlog("")
        dlog("tag src: " + self.srcName)
        dlog("this buffer: " + cBufferFN)
        checkFN = self.srcName
        if cBufferFN == self.srcName:
            # search through the modified buffer instead
            checkFN = PLUGIN_HOME + "/../tmp/cBuffers/cBuffer." + self.srcName.split(".")[-1]
            dlog("checkFN: " + checkFN)

        with open(checkFN, 'r') as f: 
            reg = self.truncate_regex()

            # why do I not have to escape these in vim? @wtodo @frequent-function @python 
            reg = reg.replace("(", "\(")
            reg = reg.replace(")", "\)")
            pat = re.compile(reg, re.MULTILINE)
            # existence of a regex in string @example @regex @python 
            if pat.search(f.read()): 
                return True
            else:
                return False

    def truncate_regex(self):

        tRegex = self.regex
        if self.regex[0] == "/":
            tRegex = self.regex[1:] 

        tRegex = tRegex.replace(r'/;"', "")
        return tRegex

    def parse_ctags_output_line(self):

        ls = self.ctagsLine.strip().split("\t")
        self.name = ls[0]
        self.srcName = ls[1]
        self.regex = ls[2]
        self.type = ls[3]
        self.lineNumber = int(ls[4].replace("line:", ""))
        try:
            self.prototype = ls[5]
        except IndexError:
            self.prototype = None 

    def to_json(self):
        # make a json dict maker @wtodo @python 
        jDict = {"name": self.name,
                 "srcName": self.srcName,
                 "regex": self.regex,
                 "type": self.type,
                 "lineNumber": self.lineNumber,
                 "prototype": self.prototype,
                 "cursorOffsetLine": self.cursorOffsetLine,
                 "cursorColumnNumber": self.cursorColumnNumber
                 }

        return json.dumps(jDict)

    def from_json(self):
       objProperties = json.loads(self.jsonLine.strip()) 
       for attName, attProp in objProperties.items():
           setattr(self, attName, attProp)

def test_tagify(srcFileName):
    '''just testing if ctagging works'''
    tFile = TagFile(srcFileName)
    print tFile.tags[0].name

def log_tag_info(srcFile, lineNum, colNum, fileExtension, logFile):
    
    #get the nearest tag, will return None if no tag match
    cBufferFN = PLUGIN_HOME + "/../tmp/cBuffers/cBuffer." + fileExtension
    
    nTag = get_nearby_tag(cBufferFN, lineNum, colNum)
  
    if nTag:
        #change the file name to not point to the temp file
        nTag.srcName = srcFile
        with open(logFile, 'a') as f:
            f.write(nTag.to_json() + "\n")
    else:
        print "no tags"

def get_nearby_tag(srcFile, lineNum, colNum, tagTypes = "fm"):
    ''' f = function, m = method '''
    lineNum, colNum = int(lineNum), int(colNum)
    tFile = TagFile(srcFile)
    lineNum_tag = cgtools.groupby(tFile.tags, "lineNumber")

    for i in xrange(lineNum, 0, -1):
        if i in lineNum_tag:
            nearTag = lineNum_tag[i][0]
            if nearTag.type in tagTypes:
                nearTag.cursorOffsetLine = lineNum - i
                nearTag.cursorColumnNumber = colNum
                return nearTag 

    # no tag was found
    return None

def update_log(logFN, cBufferFN):
    '''remove non-existing tags (file changed/deleted or refactored)
    and remove duplicate functions'''
    
    # re-factor log to class @wtodo @frequent-function
    tags = get_unique_tags(load_tags_from_log(logFN))

    # does the file even exist anymore?
    tags = [tag for tag in tags if os.path.exists(tag.srcName)]

    # debug level @programming @wtodo

    # does the function exist anymore, has it been refactored?
    tags = [tag for tag in tags if tag.check_existence(cBufferFN)] 

    with open(logFN, 'w') as f:
        [f.write(tag.to_json() + "\n") for tag in tags]

def get_unique_tags(tags):
    '''tags are returned in order given
    [a,b,c] -> [a,b] (if c was not unique)'''

    uniqueProps = set()
    uniqueTags = []
    while True:    
        if not tags:
            break

        currTag = tags.pop()
        tagUniqueProps = (currTag.name, currTag.srcName, currTag.prototype)
        if tagUniqueProps not in uniqueProps:
            uniqueTags.append(currTag)
            uniqueProps.add(tagUniqueProps)

    # popping and appending reversed the list order
    uniqueTags.reverse()

    return uniqueTags

def load_tags_from_log(logFN):

    tags = []
    with open(logFN, 'r') as f:
        for jLine in f: 
            tags.append(Tag(jsonLine = jLine))

    return tags

def generate_mru_browser_text(cBufferName):

    #remove dups and non-existing first
    update_log(LOG_FILE, cBufferName)

    #get resulting tags and display in reverse order (log has newest at end)
    logTags = load_tags_from_log(LOG_FILE)
    logTags.reverse()
    
    with open(WINDOW_TEXT, 'w') as f:
        for tag in logTags:
            if tag.prototype:
                tName = tag.name + "(" + tag.prototype.replace("class:", "") + ")"
            else:
                tName = tag.name

            outText = tName + "\t" + os.path.basename(tag.srcName) + "\t" + str(tag.cursorOffsetLine) + "\t" + str(tag.cursorColumnNumber) + "\t" + tag.regex.replace(r'/;"', "") + "\t" + tag.srcName
            f.write(outText + "\n")

if __name__ == "__main__":
    import sys

    event = sys.argv[1]
    if event == "log":
        log_tag_info(sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], LOG_FILE)
    elif event == "browsertext":
        dlog(sys.argv)
        generate_mru_browser_text(sys.argv[2])
    else:
        raise KeyError("Must Specify Event")
     


plugin/pythontag.py
import re
import logging

from mruutility import first_non_whitespace_index, is_comment_line, is_blank_line

def test_near_tag():
    with open("./tests/mockFile1.py", 'r') as f:
        fileText = f.read()

    nearestFunc = py_get_nearest_tag(fileText, 1, 0)
    assert nearestFunc == None
    
    nearestFunc = py_get_nearest_tag(fileText, 3, 5)
    assert nearestFunc[1] == "def func_outside():"

    nearestFunc = py_get_nearest_tag(fileText, 5, 8)
    assert nearestFunc[1] == "def func_outside():"

    nearestFunc = py_get_nearest_tag(fileText, 10, 10)
    assert nearestFunc[1] == "def func_inside():"
     
    nearestFunc = py_get_nearest_tag(fileText, 16, 8)
    assert nearestFunc[1] == "def func_outside():"
    
    nearestFunc = py_get_nearest_tag(fileText, 17, 0)
    assert nearestFunc == None
    
def py_get_nearest_tag(fileText, lineNum, byteNum):
    '''
    cursor is in tag's scope.  Which tag?
    lineNum input is 1-based
    byteNum input is 0-based
    '''

    #get lines but preserve newlines
    fileLines = fileText.split("\n") 
    fileLines = ["%s\n" % x for x in fileLines]

    fxnDefPattern = re.compile(r'^(\s*)(def\s+\S+)\(', re.MULTILINE)

    currLine = lineNum - 1 
    currByte = byteNum
    while True:
        
        #get left most non-whitespace character
        #don't change the currByte if only whitespace or comments
        lineText = fileLines[currLine]
        logging.debug("l %s b %s", currLine, currByte)
        logging.debug("currentPosition [%s]", lineText[currByte:])
        
        #  maybe update tag object to know it's column definition position
        #move to the next line
        logging.debug( "checking for function")
        m = fxnDefPattern.match(lineText)
        if m:
            if m.start(2) < currByte: 
                logging.debug( m.group(2))
                return (currLine + 1, m.group(2))

        #check if non-coding line (blank or comment-only line)
        logging.debug("checking non-coding")
        if is_comment_line(lineText) or is_blank_line(lineText):
            logging.debug("  encountered non-coding line")
            if currLine == 0:
                return None
            currLine -= 1
            continue
        logging.debug(" not non-coding")

        #move to first coding character on line
        #  only if fnwi is less indented than the current position
        fnwi = first_non_whitespace_index(lineText)
        currByte = fnwi if (fnwi < currByte) else currByte
        if currByte == None:
            raise ValueError("Can't find coding byte on line")
        
        logging.debug(" fnwi [%s]",  lineText[currByte:])
        
        if currLine == 0:
            return None

        currLine -= 1 

if __name__ == "__main__":
    import sys
    assert sys.argv[1] in globals(), "Need name of fxn to run from command line!"
    fxnToRun = globals()[sys.argv[1]] 
    fxnToRun(*sys.argv[2:])
